golang/cosmos/x/swingset/keeper/iterator_helpers.go

```diff
@@ -0,0 +1,375 @@
+		// Workaround for cosmos iterators not following tendermint iterator semantics
+		// Since cosmos iterators do not use Error() to communicate iteration errors,
+		// panic if we encounter an iteration error
+		panic(iter.err)
+	}
+	return true
+}
+
+// checkValid implements the Iterator validity invariants of public methods
+func (iter *exportDataIterator) checkValid() {
+	if !iter.Valid() {
+		panic("invalid iterator")
+	}
+}
+
+// Next moves the iterator to the next entry from the reader.
+// If Valid returns false, this method will panic.
+func (iter *exportDataIterator) Next() {
+	iter.checkValid()
+
+	iter.current, iter.err = iter.reader.Read()
+
+	if iter.err == nil && (len(iter.current) < 1 || len(iter.current) > 2) {
+		iter.err = fmt.Errorf("invalid data entry (length %d)", len(iter.current))
+	}
+}
+
+// Key returns the key at the current position. Panics if the iterator is invalid.
+// CONTRACT: key readonly []byte
+func (iter *exportDataIterator) Key() (key []byte) {
+	iter.checkValid()
+
+	return []byte(iter.current[0])
+}
+
+// Value returns the value at the current position. Panics if the iterator is invalid.
+// CONTRACT: value readonly []byte
+func (iter *exportDataIterator) Value() (value []byte) {
+	iter.checkValid()
+
+	if len(iter.current) == 1 {
+		return nil
+	} else {
+		return []byte(iter.current[1])
+	}
+}
+
+// Error returns the last error encountered by the iterator, if any.
+func (iter *exportDataIterator) Error() error {
+	err := iter.err
+	if err == io.EOF {
+		return nil
+	}
+
+	return err
+}
+
+// Close closes the iterator, releasing any allocated resources.
+func (iter *exportDataIterator) Close() error {
+	return iter.reader.Close()
+}
+
+// exportDataEntriesSource represents a random access source used by the
+// exportDataEntriesReader. It abstracts over any random access structure
+// which holds key/value pairs.
+type exportDataEntriesSource interface {
+	// GetAt should return the key/value pair at a given position in the backing structure
+	GetAt(index int) (ExportData, error)
+}
+
+var _ ExportDataReader = &exportDataEntriesReader{}
+
+// exportDataEntriesReader is the ExportDataReader using an
+// underlying exportDataEntriesSource. It reads from the source sequentially,
+// performs bounds checks, and tracks the current position.
+type exportDataEntriesReader struct {
+	source  exportDataEntriesSource
+	length  int
+	current int
+}
+
+// newExportDataEntriesReader returns an ExportDataReader backed by
+// a random access exportDataEntriesSource.
+func newExportDataEntriesReader(source exportDataEntriesSource, sourceLength int) ExportDataReader {
+	return &exportDataEntriesReader{
+		source:  source,
+		length:  sourceLength,
+		current: 0,
+	}
+}
+
+// Read yields the next "export data" from the source
+// Implements ExportDataReader
+func (reader *exportDataEntriesReader) Read() (next ExportData, err error) {
+	if reader.source == nil {
+		return next, fmt.Errorf("reader closed")
```

Comment:
Nit: I find use of explicit `nil` more comprehensible than use of an implicitly zero-valued named return parameter.
```suggestion
		return nil, fmt.Errorf("reader closed")
```
Which would require updating the return signature from `(ExportData, error)` to `(*ExportData, error)`. I leave the decision up to you.

Reply:
Oh this is actually a leftover from the array to slice conversion!